File: Computer Science 130/lab 10 - Stacks/q3.py
class Stack():
    def __init__(self):
        self.items = []
    def push(self, item):
        self.items.append(item)
    def pop(self):
        return self.items.pop(-1)
        
s = Stack()
s.push(5)
s.push(4)
s.push(9)
n1 = s.pop()
n2 = s.pop()
s.push(n2 * n1)
s.push(3)
n1 = s.pop()
n2 = s.pop()
s.push(n2 + n1)
n1 = s.pop()
n2 = s.pop()
s.push(n2 - n1)

print(s.items)

File: leetcode/q3.js
// 3. Longest Substring Without Repeating Characters
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  /**
   * logic:
   * Essentially keep track of the maximum size of the stack and clear all the way until the
   * character repeats again.
   *
   * [a, b, c] - > 3
   * another occurence of a!
   * a, [b, c, a] (still 3)
   */
  let longestLength = 0;
  let stack = [];
  for (let char of s) {
    // loop through the string
    if (stack.includes(char)) {
      // when the stack already has the character, we need to pop until
      stack.splice(0, stack.lastIndexOf(char) + 1);
      stack.push(char);
    } else {
      stack.push(char);
      if (stack.length > longestLength) {
        longestLength = stack.length;
      }
    }
  }
  return longestLength;
};
